sat/cli/bootsys/cabinet_power.py

```diff
@@ -24,19 +24,110 @@
 """
 Powers on and off liquid-cooled compute cabinets.
 """
+import copy
+import datetime
 import logging
 
+import croniter
+from csm_api_client.k8s import load_kube_api
+from dateutil.tz import tzutc
+from kubernetes.client import ApiException, BatchV1Api, V1CronJob
+from kubernetes.config import ConfigException
+
 from sat.apiclient import APIError, CAPMCClient, HSMClient
 from sat.cli.bootsys.power import CAPMCPowerWaiter
 from sat.config import get_config_value
-from sat.hms_discovery import HMSDiscoveryCronJob, HMSDiscoveryError, HMSDiscoveryScheduledWaiter
+from sat.hms_discovery import (HMSDiscoveryCronJob, HMSDiscoveryError,
+                               HMSDiscoveryScheduledWaiter)
 from sat.session import SATSession
 from sat.util import prompt_continue
 
-
 LOGGER = logging.getLogger(__name__)
 
 
+def cronjob_stuck(cronjob, curr_time=None):
+    """Determine if a cronjob is not being scheduled.
+
+    A cronjob is considered "stuck" if it is not suspended, but its last
+    scheduled time is before the expected most recent scheduled time,
+    based on its cron schedule. If the job is suspended, we expect it not
+    to be scheduled, so we don't treat it as "stuck".
+
+    Args:
+        cronjob (V1CronJob): the cron job object from the Kubernetes API
+        curr_time (datetime): if not None, use this as the current time.
+            Mostly used for testing.
+
+    Returns:
+        Whether the cronjob's last scheduled job was before the
+        calculated most recent scheuled time based on the cron
+        schedule string for the cron job.
+    """
+    if cronjob.spec.suspend:
+        return False
+    it = croniter.croniter(cronjob.spec.schedule)
+    if curr_time is not None:
+        it.set_current(curr_time)
+    if cronjob.status.last_schedule_time is None:
+        last_sched = cronjob.metadata.creation_timestamp
+    else:
+        last_sched = cronjob.status.last_schedule_time
+    prev_expected_sched = datetime.datetime.fromtimestamp(it.get_prev(), tz=tzutc())
+    return last_sched < prev_expected_sched
+
+
+def recreate_cronjob(batch_api, cronjob):
+    """Delete and recreate a cronjob in Kubernetes.
+
+    Args:
+        batch_api (kubernetes.client.BatchV1Api): the Kubernetes API client
+        cronjob (V1CronJob): the cronjob object representing the cronjob to be recreated
+
+    Returns:
+        V1CronJob: the newly created cronjob
+
+    Raises:
+        kubernetes.client.ApiException: if a problem occurs while communicating
+            with the Kubernetes API
+    """
+    cronjob = copy.deepcopy(cronjob)
+    name = cronjob.metadata.name
+    namespace = cronjob.metadata.namespace
+    cronjob.metadata.resource_version = None
+    try:
+        batch_api.delete_namespaced_cron_job(name, namespace)
+    except ApiException as err:
+        if err.status != 404:
```

Comment:
This is sort of an artifact of my testing, as I was initially passing in the name and namespace as arguments to `recreate_cronjob()` instead of the `V1CronJob` object. I think it still can serve a purpose, since IMO if we "recreate" a cronjob that doesn't exist, the logical end result would be that the cron job is created after all.

I also wanted to be a bit defensive here since the recreation operation isn't atomic; if for any reason this function fails between deleting and recreating the cronjob, then there could be a situation where _no_ cron job exists, and I wanted to smooth over an easy-to-handle error case.

Reply:
That makes sense, thanks.